numpy/core/src/multiarray/descriptor.c

```diff
@@ -90,31 +90,51 @@ _arraydescr_from_ctypes_type(PyTypeObject *type)
  * and it can be converted to a dtype object.
  *
  * Returns a new reference to a dtype object, or NULL
- * if this is not possible. When it returns NULL, it does
- * not set a Python exception.
+ * if this is not possible.
+ * When the return value is true, the dtype attribute should have been used
+ * and parsed. Currently the only failure mode for a 1 return is a
+ * RecursionError and the descriptor is set to NULL.
+ * When the return value is false, no error will be set.
  */
-NPY_NO_EXPORT PyArray_Descr *
-_arraydescr_from_dtype_attr(PyObject *obj)
+int
+_arraydescr_from_dtype_attr(PyObject *obj, PyArray_Descr **newdescr)
 {
     PyObject *dtypedescr;
-    PyArray_Descr *newdescr = NULL;
     int ret;
 
     /* For arbitrary objects that have a "dtype" attribute */
     dtypedescr = PyObject_GetAttrString(obj, "dtype");
-    PyErr_Clear();
     if (dtypedescr == NULL) {
-        return NULL;
+        /*
+         * This can be reached due to recursion limit being hit while fetching
+         * the attribute (tested for py3.7). This removes the custom message.
+         */
+        goto fail;
+    }
+
+    if (Py_EnterRecursiveCall(
+            " while trying to convert the given data type from its "
+            "`.dtype` attribute.") != 0) {
+        return 1;
     }
 
-    ret = PyArray_DescrConverter(dtypedescr, &newdescr);
```

Comment:
Can we maybe deprecate `.dtype` attributes which are not themselves dtypes?

Honestly, overloading `dtype` to mean "get the dtype of this thing" already seems like a risky thing to do.
The best argument I can see for allowing it is it matches the semantics of `type`, which can be used to both inspect and construct dtypes.

However, I can't see a use case for allowing it to inspect and construct at the same time - that seems like a disaster, which results in this recursion.

Reply:
Deprecation sounds like a good idea. I am not actually sure what the real use case here is, so I am wondering if we might be able to go the extra mile and just deprecate the whole logic. Will have a look at it.

Reply:
So `np.format_parser` and record arrays require the dtype lookup. But I guess it actually makes sense to support in any case (in a way it is similar to supporting `np.float64`). f2py has some issues with the deprecation, but I am sure I can patch that.

Reply:
> and record arrays require the dtype lookup.

Mind linking to where this happens?

Reply:
Format parser is documented to work in the form of: ``dtype = format_parser(formats, names, titles).dtype``, which means passing it directly as a dtype works through attribute lookup and is inviting.

Hmmpf, will hunt down the other thing later.